src/adhocracy_core/adhocracy_core/scripts/auto_transition_process_workflow.py

```diff
@@ -0,0 +1,105 @@
+"""Automatically transition workflow states.
+
+This is registered as console script.
+"""
+import transaction
+import argparse
+import inspect
+import logging
+
+from pyramid.paster import bootstrap
+from pyramid.registry import Registry
+from substanced.util import find_service
+
+from adhocracy_core.authorization import create_fake_god_request
+from adhocracy_core.interfaces import IResource
+from adhocracy_core.interfaces import search_query
+from adhocracy_core.resources.process import IProcess
+from adhocracy_core.sheets.workflow import IWorkflowAssignment
+from adhocracy_core.workflows import transition_to_states
+from adhocracy_core.utils import now
+
+
+logger = logging.getLogger(__name__)
+
+
+def auto_transition_process_workflow():  # pragma: no cover
+    """Automatically transition workflow states.
+
+    usage::
+
+        bin/auto_auto_transition_process_workflow etc/development.ini
+    """
+    docstring = inspect.getdoc(auto_transition_process_workflow)
+    parser = argparse.ArgumentParser(description=docstring)
+    parser.add_argument('ini_file',
+                        help='path to the adhocracy backend ini file')
+    args = parser.parse_args()
+    env = bootstrap(args.ini_file)
+    _auto_transition_process_workflow(env['root'], env['registry'])
+    transaction.commit()
+    env['closer']()
+
+
+def _auto_transition_process_workflow(context: IResource, registry: Registry):
+    catalogs = find_service(context, 'catalogs')
+    query = search_query._replace(interfaces=IProcess,
+                                  resolve=True
+                                  )
+    processes = catalogs.search(query).elements
+    for process in processes:
+        workflow_assignment = registry.content.get_sheet(process,
+                                                         IWorkflowAssignment)
+        if _workflow_auto_transition_enabled(registry, workflow_assignment) \
+            and _workflow_auto_transition_needed(process, registry,
+                                                 workflow_assignment):
+            _do_workflow_auto_transition(process, registry)
+
+
+def _workflow_auto_transition_enabled(
+        registry: Registry, workflow_assignment: IWorkflowAssignment):
+    workflow_name = workflow_assignment.get()['workflow']
+    workflow_meta = registry.content.workflows_meta[workflow_name]
+    return workflow_meta['auto_transition']
+
+
+def _workflow_auto_transition_needed(context: IResource, registry: Registry,
+                                     workflow_assignment: IWorkflowAssignment):
+    workflow = registry.content.get_workflow(context)
+    next_states = workflow.get_next_states(context,
+                                           create_fake_god_request(registry))
+    if len(next_states) == 1:
+        next_state = next_states[0]
+        end_date = _get_current_state_end_date(workflow_assignment, next_state)
+        if end_date:
+            return now() > end_date
```

Comment:
it would be better to have this measured once when the script begin. Otherwise it could have some strange effects, for example workflow w1 and workflow w2 have and end date of t1 and t2 respectively. Now the script is executed at time t1+d where t1+d < t2. The following could occur: updating workflow w1, for any reason, take e time where t1+d+e > t2. When this check occurs, `now()` will be superior or equal to t1+d+e and t2 will also be updated, which breaks the principle of least astonishment: the user would have expected only t1 to be updated.

I know the example seems a bit contrived but I prefer to be on the safe side.